### PR TITLE
[FIX] hr_attendance: user can see other users attendances

### PR DESCRIPTION
**Pull request by backspac**
Steps:
- As admin, go to Settings > Users & Companies > Users
- Edit Mark Demo (demo)
- In Human Resources > Attendances, select Manual Attendance or blank
- As demo, go to My Profile
- Click the smart button showing the hours worked for the last month
- Remove all filters

Bug:
The demo user, who hasn't the rights to see the other employees
attendances, can see them.

Explanation:
Every user must have the right to read attendances in order to see their
own attendances. Not giving the users the read rights in the security
record rule prevents the record rule from being applied when reading
attendances. This makes the read access rights the only rule and allows
everyone to see the attendances of the others.

This commit also fixes the default selected employee when going to the
attendances tree view on these paths:
- User
- Employee
- User > Employee

In fact, sometime, `active_id` is the ÌD of the user and not of the
employee. This leads to incorrect results since another employee's
attendances are shown.

Finally, this commit prevents users from creating attendances from other
apps since only attendance officers and above can have access to the
creation form within the Attendances app.

opw:2440117

### CODE REVIEW

**Comment by backspac** [comment]
Ping @nle-odoo

**Review by nle-odoo — COMMENTED** [review]
so with this change, you will only see:

hr.employee.base().last_attendance_id
hr.employee.base().attendance_ids
hr.employee.base().last_check_in
hr.employee.base().last_check_out
res.users().last_check_in
res.users().last_check_out

of yourself.

I didn't find anywhere where that could cause issue in stable, from what I found we use these only for attendance officers and elsewhere we use attendance_state which uses sudo.

So maybe we should check with odo or mat to see if they see an issue for stable (for 13.0 this would not affect too much people since it only applies on new installation of hr_attendance).

With this change:

```diff
-        <field name="context">{'search_default_employee_id': active_id, 'default_employee_id': active_id, 'create': False}</field>
+        <field name="context">{'create': False}</field>
```

clicking on one attendance of one employee (you need to be hr officer):

```
                <button name="%(hr_attendance_action_employee)d"
                    id="hr_attendance_button"
                    class="oe_stat_button"
                    type="action" attrs="{'invisible': [
                        '|', ('attendance_state', '=', False),
                        '&amp;',
                            ('hr_presence_state', '=', 'present'),
                            ('attendance_state', '=', 'checked_out')]}">
```

 display all the employees.

In:

```diff
- context="{'search_default_employee_id': active_id, 'search_default_check_in_filter': '1'}"
+ context="{'search_default_employee_id': id, 'search_default_check_in_filter': '1'}"
```

why is active_id replaced? it doesn't work?

also why changing the action there:

```diff
-                <button name="%(hr_attendance_action)d"
+                <button name="%(hr_attendance_action_employee)d"
                        class="oe_stat_button"
                        icon="fa-calendar"
                        type="action"
                        context="{'search_default_employee_id': employee_ids, 'search_default_check_in_filter': '1'}"
                        groups="base.group_user"
                        help="Worked hours last month">
```

Trying with or without your code, it seems to show me the same thing (a view for last month filtering the employee I clicked on).

**Comment by backspac** [comment]
`id` instead of `active_id` is because, sometimes, `active_id` is the user ID and not the employee ID. For example, when you go from Users to Employees:
user

For me, `hr_attendance_action_employee` should be used in the employee/user views and `hr_attendance_action` in the Attendances module. The only notable difference is the kanban view which is present in `hr_attendance_action`. It also prevents creating new entries.

**Review by nle-odoo — APPROVED** [review]
> `id` instead of `active_id` is because, sometimes, `active_id` is the user ID and not the employee ID. For example, when you go from Users to Employees

yeah looks good then

> For me, hr_attendance_action_employee should be used in the employee/user views and hr_attendance_action in the Attendances module. The only notable difference is the kanban view which is present in hr_attendance_action. It also prevents creating new entries.

okay. then you should ensure in each cases you still see the same thing:

- when having a custom action using hr_attendance_action_employee it still work as before
- where we used hr_attendance_action and search_default_today flter worked that it still works

if yes it ok to merge I think

robodoo delegate+ 

edit: there is still the issue of changing access rights that I forgot about, I'm not sure if it's ok or not in stable:

> I didn't find anywhere where that could cause issue in stable, from what I found we use these only for attendance officers and elsewhere we use attendance_state which uses sudo.
> 
> So maybe we should check with odo or mat to see if they see an issue for stable (for 13.0 this would not affect too much people since it only applies on new installation of hr_attendance).

**Comment by backspac** [comment]
> I didn't find anywhere where that could cause issue in stable, from what I found we use these only for attendance officers and elsewhere we use attendance_state which uses sudo.
>
> So maybe we should check with odo or mat to see if they see an issue for stable (for 13.0 this would not affect too much people since it only applies on new installation of hr_attendance).

@odony @mart-e What do you think about this?

File: addons/hr_attendance/security/hr_attendance_security.xml [hunk 1]
```diff
@@ -62,7 +62,7 @@
             <field name="name">user: modify own attendance only</field>
             <field name="model_id" ref="model_hr_attendance"/>
             <field name="domain_force">[('employee_id.user_id','=',user.id)]</field>
-            <field name="perm_read" eval="0"/>
+            <field name="perm_read" eval="1"/>
```
Comment (odony) [inline addons/hr_attendance/security/hr_attendance_security.xml:65]: From a stability standpoint, it's ok to change this in stable, as it will only affect new databases.

From a privacy point of view, I think it's better to prevent `read` indeed, but as you can see with rule name, this was not the intended behavior initially. So you'll need to update the rule name too, and confirm this change functionally with a PO first.

Remember, we don't make changes in business logic just because someone opens a ticket, especially when there is no traceback/error. Because each customer has a different opinion on "the expected behavior", we need take our own decision, and stick to it  ;-)

File: addons/hr_attendance/views/hr_attendance_view.xml [hunk 2]
```diff
@@ -148,7 +148,7 @@
         </field>
     </record>
 
-    <record id="hr_attendance_action_employee" model="ir.actions.act_window">
+    <record id="hr_attendance_action_employee" model="ir.actions.act_window"> <!-- Not used -->
```
Comment (odony) [inline addons/hr_attendance/views/hr_attendance_view.xml:151]: Why mark this record unused and make a copy? You could just modify the original action and reduce the diff. It won't affect databases that do not update the `hr_attendance` module, and when they do, everything will be updated.

File: addons/hr_attendance/security/hr_attendance_security.xml (hunk 1 above)
Comment (backspac) [reply addons/hr_attendance/security/hr_attendance_security.xml:65]: Thanks! I've changed the name.

> confirm this change functionally with a PO first

Indeed. FLL was OK with the change.

> Remember, we don't make changes in business logic just because someone opens a ticket, especially when there is no traceback/error. Because each customer has a different opinion on "the expected behavior", we need take our own decision, and stick to it ;-)

:+1:

File: addons/hr_attendance/views/hr_attendance_view.xml (hunk 2 above)
Comment (backspac) [reply addons/hr_attendance/views/hr_attendance_view.xml:151]: Because a customization using this view would be impacted by these changes. I've tried to override the view's context in the button, but it doesn't work that way.

Comment (nle-odoo) [reply addons/hr_attendance/views/hr_attendance_view.xml:151]: I preferred not to have view used only for attendance office be used for seeing your own attendance (my profile -> attendance) and that other people that used the view expecting it to filter on active_id don't get it non-filtered all of a sudden.

If it's okay we could change `hr_attendance_action_employee` as done initially.

I'm not sure the change in view_employee_form_inherit_hr_attendance:

```diff
-                 <button name="%(hr_attendance_action)d"
+                <button name="%(hr_attendance_action_no_create)d"
```

is necessary since we only see that button when we are an attendance officer.

Comment (odony) [reply addons/hr_attendance/views/hr_attendance_view.xml:151]: > Because a customization using this view would be impacted by these changes. I've tried to override the view's context in the button, but it doesn't work that way.

It's great that you're trying to avoid break customisations, but in this case I find it overkill:

1) The chances of having a customization impacted by this change are very small, because the change is tiny (just the `context` AFAICS). What could break? It's very hard for a customisation to hard-depend on that. The behavior will be sightly different, but that's the whole point of the fix.
2) The change won't impact any database unless they force-update the module, or go through an upgrade. At that point, so many things can break in terms of customisations, and this one is a very minor concern.
3) By introducing a new XML ID, you *also* add some risk: for a future bugfix in another module, say e.g. `hr_holidays_attendance`, you may be tempted to use this new XML ID. But now if someone installs module `hr_holidays_attendance` while they have never updated `hr_attendance`, it will crash (unknown XML ID).
4) Finally, by duplicating the action you're introducing some tech debt in the codebase, as someone will have to cleanup the dead code later (or it will be forgotten).

Long story short, as with all things, common sense prevails, even with the stable policy. Think in terms of the number of customer databases that will be possibly affected by the change (here: I can hardly imagine a single one), and the severity of the problem they will experience if that ever happens (here: very minor behavior change), and choose the option with the best risk/benefit ratio :)

**Comment by backspac** [comment]
@robodoo r+
